vpr/src/place/place_constraints.cpp

```diff
@@ -257,3 +258,149 @@ void load_cluster_constraints() {
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+    auto& place_ctx = g_vpr_ctx.mutable_placement();
+    auto& floorplanning_ctx = g_vpr_ctx.floorplanning();
+
+    for (auto blk_id : cluster_ctx.clb_nlist.blocks()) {
+        if (!is_cluster_constrained(blk_id)) {
+            continue;
+        }
+        PartitionRegion pr = floorplanning_ctx.cluster_constraints[blk_id];
+        auto block_type = cluster_ctx.clb_nlist.block_type(blk_id);
+        t_pl_loc loc;
+
+        if (is_pr_size_one(pr, block_type, loc)) {
+            //Set block location and grid usage
+            set_block_location(blk_id, loc);
+
+            //Set as fixed
+            place_ctx.block_locs[blk_id].is_fixed = true;
+        }
+    }
+}
+
+/*
+ * Returns 0, 1, or 2 depending on the number of tiles covered.
+ * Will not return a value above 2 because as soon as num_tiles is above 1,
+ * it is known that the block that is assigned to this region will not be fixed, and so
+ * num_tiles is immediately returned.
+ * Updates the location passed in because if num_tiles turns out to be 1 after checking the
+ * region, the location that was set will be used as the location to which the block
+ * will be fixed.
+ */
+int region_tile_cover(const Region& reg, t_logical_block_type_ptr block_type, t_pl_loc& loc) {
+    auto& device_ctx = g_vpr_ctx.device();
+    vtr::Rect<int> rb = reg.get_region_rect();
+    int num_tiles = 0;
+
+    for (int x = rb.xmin(); x <= rb.xmax(); x++) {
+        for (int y = rb.ymin(); y <= rb.ymax(); y++) {
+            auto& tile = device_ctx.grid[x][y].type;
+
+            /*
+             * If the tile at the grid location is not compatible with the cluster block
+             * type, do not count this tile for num_tiles
+             */
+            if (!is_tile_compatible(tile, block_type)) {
+                continue;
+            }
+
+            /*
+             * If the region passed has a specific subtile set, increment
+             * the number of tiles set the location using the x, y, subtile
+             * values
+             */
+            if (reg.get_sub_tile() != NO_SUBTILE) {
+                num_tiles++;
+                loc.x = x;
+                loc.y = y;
+                loc.sub_tile = reg.get_sub_tile();
+                if (num_tiles > 1) {
+                    return num_tiles;
+                }
+
+                /*
+                 * If the region passed does not have a subtile set, set the
+                 * subtile to zero. The loc that is set will only be used in the
+                 * event that num_tiles is 1, and num_tiles will only be 1 if the
+                 * capacity of the tile type turns out to be zero, thus implying
+                 * that the location subtile will be zero.
+                 */
+            } else if (reg.get_sub_tile() == NO_SUBTILE) {
+                auto& cap = tile->capacity;
+                for (int z = 0; z < cap; z++) {
+                    num_tiles++;
```

Comment:
Again, don't we have to check if *this* subtile (z) is compatible in type, and only increment num_tiles if it is?  And then we should set sub_tile to the first z value that passes that test.